Avoid reloading history if crafting window is closed

clery opened the pull request:
In addition to your proposed changes, here are a few changes on what we talked about in [#148](https://github.com/derfloh205/CraftSim/pull/148)

It's not perfect by any means but it should alleviate some of the issues you've talked about, like the lags induced by reloading the history view even though it's not in use.
Don't hesitate to tell me if you see anything weird :)

(I took the liberty to do the pull request on your fork to avoid merging issues later on)

Commit:
Fix max history size not being saved

Commit:
Reload history only when crafting window is opened

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -58,11 +58,10 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
 local function GetFallbackCustomer()
+    if CraftSim.CUSTOMER_HISTORY.FRAMES.lastCustomer then
+        return CraftSim.CUSTOMER_HISTORY.FRAMES.lastCustomer
+    end
```
I don't think that we still need the loop with this? We could probably get rid of that method and simply do this in `SetCustomer`:

```lua
customer = customer or CraftSim.CUSTOMER_HISTORY.FRAMES.lastCustomer
if not customer then
    return
end
```

clery replied:
I think we still need it, when we open the customer history but haven’t received any messages beforehand. Then we do need to get the first customer from the history map

clery replied:
Unless we would rather just show an empty history view and let the user select a specific customer before showing any history

Meivyn replied:
We do actually need it indeed, because there's no logic to load the history otherwise. If you remove that part then the history and the dropdown will always be empty on `/reload` or login. Either keep the loop or make this a saved setting so it does at least populate the dropdown and load the history for that specific customer.

The point of the loop is only to skip the `version` key honestly. So it's up to you what behavior you prefer.

clery replied:
Keeping the last selected customer through reload and login isn't a bad idea. Let me think on it for a bit !

Commit:
Add the ability for the customer history to remember last customer selected

Commit:
Add the ability to remove a customer

Commit:
Add filter text input to search more easily through customers. Change migration to remove customers without tips

clery commented:
So I added a couple things based on your feedback : 
![image](https://github.com/Meivyn/CraftSim/assets/7500492/81921f80-837e-4793-85d3-8b4dfdb24eb6)
Here is a search box. You type in something and it searches through customers (regexp compatible)

![image](https://github.com/Meivyn/CraftSim/assets/7500492/17066e30-7bb2-450e-9f79-83cdb6577911)
You can also see that I added a much needed `Remove customer` button. That's pretty self explanatory.

And finally I opted into saving the last customer selected to reopen his history on reload/login. That way if you get whispered on an alt or something, changing characters, you'll still be on the right history

Let me know if you see any issues.
One known issue is : if you search with an invalid regexp, it throws a Lua error. I couldn't really find how to properly handle the error, so I thought it wasn't that big of a deal, but if you have more experience with Lua, you're welcome to make adjustments !

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -24,15 +24,46 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
+        "",
+        function(textInput, userInput)
+            if userInput then
+                input = textInput:GetText() or ""
```
This should be `local`.

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -58,15 +89,9 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
-            return k
-        end
+    if CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer then
+        return CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer
```
Again I see no point in keeping the `GetFallbackCustomer` method with this change. While negligible, it adds an extra method call and an extra table lookup (and extra verbosity).

This also will still return `nil` if the `lastCustomer` has been removed with the "Remove customer" button, but that's not really a big deal I guess.

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -24,15 +24,46 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
+                input = textInput:GetText() or ""
+                self:ResetDropdown(CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer, input)
+                if not _G["DropDownList"..1]:IsShown() then
+                    ToggleDropDownMenu(1, nil, self.frame.content.customerDropdown.frame, self.frame.content.customerDropdown.frame.name, 170, 50, nil, nil)
```
I don't think you want to call this if `input` is an empty string?

Meivyn commented on Modules/CustomerHistory/CustomerHistory.lua:
```diff
@@ -34,13 +34,14 @@ function CraftSim.CUSTOMER_HISTORY:HandleWhisper(event, message, customer, ...)
-    self:UnregisterEvent("TRADE_SKILL_SHOW")
+function CraftSim.CUSTOMER_HISTORY:LoadHistory()
+    if not self.db.realm.lastCustomer then
+        CraftSim.CUSTOMER_HISTORY.FRAMES:ResetDropdown()
```
You still call this twice if `lastCustomer` is `nil`.

Meivyn commented on Modules/CustomerHistory/CustomerHistory.lua:
```diff
@@ -34,13 +34,14 @@ function CraftSim.CUSTOMER_HISTORY:HandleWhisper(event, message, customer, ...)
-    self:UnregisterEvent("TRADE_SKILL_SHOW")
+function CraftSim.CUSTOMER_HISTORY:LoadHistory()
+    if not self.db.realm.lastCustomer then
+        CraftSim.CUSTOMER_HISTORY.FRAMES:ResetDropdown()
```
I am not sure to understand why this call is needed, mind explaining? The dropdown should already be initialized with empty data:

https://github.com/Meivyn/CraftSim/blob/9b0dd555896910978e2cd627c89c0aad45f9d12d/Modules/CustomerHistory/Frames.lua#L28-L31

clery replied:
So what happened was : let's say I /reload and `lastCustomer` is nil, when opening the tradeskill window, the dropdown would be empty because calling `SetCustomer` with a nil customer returns, and thus does not reload the dropdown.

But you're right, this is dumb. I'll just move the return condition in SetCustomer to LoadHistory because it's the only function where customer can't be nil

clery replied:
Also, ResetDropdown doesn't empty the dropdown, it resets it to a list of customer (partial if a filter is provided)

Meivyn replied:
It would be empty but filled as soon as you click on it/search for something, right? Nothing would be selected anyway since `lastCustomer` is nil, and the user can't notice that it is empty. At least that's what I think?

Meivyn replied:
It is actually loaded because of `OnTextChanged` being fired when the filter box is initialized.

clery replied:
Nah that’s the thing, it wouldn’t have loaded the dropdown at all. Both nothing selected and an empty list. But it should be fine with the recent changes I made

clery replied:
> It is actually loaded because of `OnTextChanged` being fired when the filter box is initialized.

Ah yeah, but if you were to just open the drop-down it would’ve be empty. Anyway that’s all in the past now

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -24,15 +24,46 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
+                CraftSim.CUSTOMER_HISTORY.db.realm[CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer] = nil
+                self:ResetDropdown()
+                self.frame.content.messageBox:Clear()
+                self.frame.content.totalTip:SetText(string.format("%s%s", CraftSim.LOCAL:GetText(CraftSim.CONST.TEXT.CUSTOMER_HISTORY_TOTAL_TIP), CraftSim.GUTIL:FormatMoney(0)))
```
It would make more sense to call `self:LoadTotalTip()` here, by adding some logic if `customer` is `nil` instead of doing this in my opinion.

Commit:
CustomerHistory total tip is now empty if no customer is selected

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -24,15 +24,46 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
+            else
+                self:ResetDropdown(CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer)
+                if _G["DropDownList"..1]:IsShown() then
+                    ToggleDropDownMenu(1, nil, self.frame.content.customerDropdown.frame, CraftSim.CUSTOMER_HISTORY.FRAMES.frame.content.customerDropdown.frame.name, 0, 0, nil, nil)
```
If the dropdown is shown then show it? I am not sure to understand this part.

clery replied:
This is a weird trick. The only way I found to Show or Hide a dropdown was to use ToggleDropDownMenu.
The idea was that if the user input is nil or empty, it would Toggle the dropdown to actually hide it. The same way above it only toggles to show it if it is not shown already

Meivyn replied:
The `userInput` is actually a boolean, meaning that it was triggered by the user's interaction. As far as I know, this part of the condition is only triggered when the box is initialized or when you set its text manually in the code.

Commit:
Fix variable 'input' that wasn't local

Commit:
Change some logic for nil customer

Commit:
Remove GetFallbackCustomer

clery commented:
I think I fixed everything based on your review 😃
I would appreciate if you could try it out. I'm not entirely sure about the DropDown filter thingy. I had to offset it so that it would not hide the text input, but because of that there is inconsistency in its position (whether filtering or directly clicking the dropdown arrow). What do you think ?

Edit: Some pictures to illustrate the issue

Here the dropdown was shown during typing
![image](https://github.com/Meivyn/CraftSim/assets/7500492/db08a4d5-c92d-4062-a50b-ea2e34d8f0c0)

Here it was shown by clicking the arrow
![image](https://github.com/Meivyn/CraftSim/assets/7500492/cf4cd4a5-6437-466e-9a28-6b61ddc4c63d)

clery commented:
I'd like to thank you again for the time you're taking for this review and for your feedback !

Meivyn commented:
I did actually test it as much as I can, trying to break it, I'm not done yet though, and still have to test your recent changes.

I don't think that the dropdown position is a problem. And you're right that its initial position would hide the input box, which is more of a problem than its offset position. I'd keep it like this (or tweak it if you think it could be better placed elsewhere).

Overall those are good changes, the remove button actually allows you to control the list of customers you want to keep without having logic to remove players that are not actual customers. And the search box is a good thing if you have *a lot* of whispers.

Have to say that having Lua errors if the search expression isn't valid is an issue though. I didn't look at it yet.

Commit:
Add pcall encapsulation around ResetDropdown to prevent errors from invalid Regexp

clery commented:
There I added a sort of "try-catch" around the ResetDropdown function content, that should do it !

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -134,11 +134,13 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:LoadHistory(customer)
-        data=data,
-    })
+    pcall(function()
+        local data = CraftSim.GUTIL:Map(CraftSim.CUSTOMER_HISTORY.db.realm, function(a, e) if (a.history and (not filter or string.find(string.lower(e), string.lower(filter)))) then return {label=e, value=e} end end)
```
I'd put the `pcall` around `string.find` instead, since this is what it should catch.

```lua
select(2, pcall(string.find, string.lower(e), string.lower(filter)))))
```

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -24,15 +24,46 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:Init()
+        function(textInput, userInput)
+            local input = textInput:GetText():gsub("^%s*(.-)%s*$", "%1") or ""
+            if input ~= "" then
+                self:ResetDropdown(CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer, input)
```
With this new logic the dropdown list simply doesn't disappear when the filter is empty, which is a bit unnatural in my opinion.

It's probably best to do something like:

```lua
if input ~= "" then
    self:ResetDropdown(CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer, input)
    if not _G["DropDownList"..1]:IsShown() then
        ToggleDropDownMenu(1, nil, self.frame.content.customerDropdown.frame, self.frame.content.customerDropdown.frame.name, 170, 50, nil, nil)
    end
elseif _G["DropDownList"..1]:IsShown() then
    ToggleDropDownMenu(1, nil, self.frame.content.customerDropdown.frame, CraftSim.CUSTOMER_HISTORY.FRAMES.frame.content.customerDropdown.frame.name, 0, 0, nil, nil)
end
```

I think that what you intended to do instead of doing `if userInput then ... else ToggleDropDownMenu end`, isn't it?

Which would result in:

```lua
if not userInput then return end
local input = textInput:GetText():gsub("^%s*(.-)%s*$", "%1") or ""
if input ~= "" then
    self:ResetDropdown(CraftSim.CUSTOMER_HISTORY.db.realm.lastCustomer, input)
    if not _G["DropDownList"..1]:IsShown() then
        ToggleDropDownMenu(1, nil, self.frame.content.customerDropdown.frame, self.frame.content.customerDropdown.frame.name, 170, 50, nil, nil)
    end
elseif _G["DropDownList"..1]:IsShown() then
    ToggleDropDownMenu(1, nil, self.frame.content.customerDropdown.frame, self.frame.content.customerDropdown.frame.name, 170, 50, nil, nil)
end
```

clery replied:
From my tests, userInput is always true, probably because it’s always shown and enabled when using it
But I may have made a mistake, I’ll take a look later when I get home

Meivyn replied:
Basically, `userInput` would be false if you call `textInput.SetText`, and that is done once in your code when you initialize the input box.

clery replied:
Yeah I get your logic. We might want to get the ResetDropdown out though, since we would want the list to be updated to either reflect an empty filter or a real filter. Like let’s say my filter is “Meivyn”, I select all and ctrl+x, it is now empty so we have to reset the dropdown list to default full list

Meivyn replied:
Ah right, that was my initial comment, but I actually was more focused on the shown state than the dropdown content. Good catch.

clery replied:
No problem 😄 
I won’t be home for a few hours, if you’re willing to you’re more than welcome to make the change !

Commit:
Better handling of Customer History filter changed callback

Commit:
pcall only around string.find instead of entire function

Meivyn commented on Modules/CustomerHistory/Frames.lua:
```diff
@@ -96,8 +131,13 @@ function CraftSim.CUSTOMER_HISTORY.FRAMES:SetCustomer(customer)
-    local data = CraftSim.GUTIL:Map(CraftSim.CUSTOMER_HISTORY.db.realm, function(a, e) if (a.history) then return {label=e, value=e} end end)
+function CraftSim.CUSTOMER_HISTORY.FRAMES:ResetDropdown(customer, filter)
+    local data = CraftSim.GUTIL:Map(CraftSim.CUSTOMER_HISTORY.db.realm, function(a, e)
+        valid, match = pcall(string.find, string.lower(e), string.lower(filter or ""))
```
This again should be `local`. You should take the habit of starting to write your variables with that prefix.

clery replied:
Ffs I hate Lua

clery replied:
Thanks for the good catch

Commit:
Fix variable to be local in CustomerHistory ResetDropdown

Meivyn committed:
Fix indent
